data/src/main/java/org/apache/iceberg/data/TableScanIterable.java

```diff
@@ -161,7 +161,10 @@ public boolean hasNext() {
 
           if (task.residual() != null && task.residual() != Expressions.alwaysTrue()) {
             Evaluator filter = new Evaluator(projection.asStruct(), task.residual(), caseSensitive);
-            this.currentIterator = Iterables.filter(reader, filter::eval).iterator();
+            this.currentIterator = Iterables.filter(reader, record -> {
+              InternalRecordWrapper wrapperRecord = new InternalRecordWrapper(record.struct()).wrap(record);
```

Comment:
The wrapper can be reused so it should be moved outside the predicate. The predicate just needs to be this:

```java
  record -> filter.eval(wrapper.wrap(record))
```